### PR TITLE
Create SimpleRegressor class #4

### LINKED ISSUE #4: Create SimpleRegressor class
**Description of feature/enhancement**
This class should handle generating the most accurate model for for any given regression problem.

**Description of implementation**
The class should have a function called `fit` which creates the most accurate regression model for any given data set.

**Why this is a good feature/enhancement**
This will allow users to simply pass in data and be returned the most accurate model without having to worry about fine tuning the hyper-parameters themselves.

**Additional context**
n/a

### LINKED ISSUE #11: SimpleRegressor Test Cases
**Description of feature/enhancement**
Test cases for SimpleRegressor class. This should ideally test multiple algorithms that are considered as the most "accurate".

**Description of implementation**
Use datasets directly from sklearn to remove the need data wrangling.

**Why this is a good feature/enhancement**
Will be able to have an automated way to validate SimpleRegressor class.

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by eranns**
Hi 

Let's discuss here regarding the regressor class. this is the first implementation.

I think these things needs to be decided : 
* which scoring use for the gridsearchcv (currently rmse) and metrics (currently mse , mae and r2)
* Is the param grid good enough or it can be improved
* threshold and method to decide if tests are good enough - unlike in classifier where we can decide by accuracy in regression we need to measure it in another way like mse , rmse , mae and so on - can say that some tests are not reaching what I expected atm and mse can be high and not close to zero.
* fail models implementation (might solve other issue #36 )

### CODE REVIEW

File: simple_learn/regressors/simple_regressor.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                    self.failed_models.append(name)
+                    print("Model: {}, Error : {} ,".format(name, e))
+                    continue
+                end = time.time()
+                if self.metrics.get(
+                    "Training Score"
+                ) is None or -grid_clf.best_score_ < self.metrics.get("Training Score"):
+                    self.metrics["Training Score"] = -grid_clf.best_score_
+                    pred_y = grid_clf.predict(train_x)
+                    self.metrics["mae"] = mean_absolute_error(train_y, pred_y)
```
Comment (skekre98) [inline simple_learn/regressors/simple_regressor.py]: Let's write the full name of the metrics out as the key in the dictionary similar to the classifier class:
```json
"Metrics": {
        "Training Accuracy": 0.9866666666666667,
        "Jaccard Score": 0.9245283018867925,
        "F1 Score": 0.96
    }
```

File: simple_learn/regressors/simple_regressor.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                    verbose=0,
+                    n_jobs=-1,
+                    error_score="raise",
+                )
+                start = time.time()
+                try:
+                    grid_clf.fit(train_x, train_y)
+                except ValueError as e:
+                    self.failed_models.append(name)
+                    print("Model: {}, Error : {} ,".format(name, e))
```
Comment (skekre98) [inline simple_learn/regressors/simple_regressor.py]: Instead of using the `print` function. Might I recommend the python [logging](https://realpython.com/python-logging/) library. It will allow us to indicate the message as debug, info, or warning. I think using the info or warning options will suffice. WDYT?

**Comment by skekre98** [comment]
Really great work @eranns, thanks for taking up this key feature 🎉 
 
> * which scoring use for the gridsearchcv (currently rmse) and metrics (currently mse , mae and r2)

I think rmse should suffice for now. I was also thinking it may be a good idea to allow the user the option to decide the scoring metric themselves. 
Ex:
```python
# In the case of classifier
def fit(self, train_x, train_y, folds=3, scoring="accuracy"):
```
We can definitely change this in a future PR.

> * Is the param grid good enough or it can be improved

I definitely think it is good for now. It might be a good idea to get a solid workflow going before worrying about model coverage. 

> * threshold and method to decide if tests are good enough - unlike in classifier where we can decide by accuracy in regression we need to measure it in another way like mse , rmse , mae and so on - can say that some tests are not reaching what I expected atm and mse can be high and not close to zero.

Yes, I like the idea of using a threshold for error and making sure the error score is under a certain value. Do you have an idea of what you think this threshold should be after implementing the class? On a side note, you will need to add the command for the regressor test cases here for them to be included in the ci pipeline:
https://github.com/skekre98/simple_learn/blob/a5097b265d68bee1adba334fcefec10270bfe94c/scripts/custom.sh#L84-L87

> * fail models implementation (might solve other issue #36 )

I have added a review comment regarding this. Let me know what you think. I really like the idea of saving the failed models as a class attribute as well.

File: simple_learn/regressors/simple_regressor.py (hunk 1 above)
Comment (eranns) [reply simple_learn/regressors/simple_regressor.py]: Ok. added full names. I noticed rmse running twice, one in Training score and one in another metric so I changed to mse.

File: simple_learn/regressors/simple_regressor.py (hunk 2 above)
Comment (eranns) [reply simple_learn/regressors/simple_regressor.py]: Yes logs are much better I didn't have that in mind when I wrote it.
I think warning is good enough. Heres the output for the code 

`self.logger.warning(f'{name} failed, Error : {value_error}')`

```
Model DecisionTreeRegressor failed , Error : Some value(s) of y are negative which is not allowed for Poisson regression.

Model HistGradientBoostingRegressor failed , Error : loss='poisson' requires non-negative y and sum(y) > 0.
```

**Comment by eranns** [comment]
> Really great work @eranns, thanks for taking up this key feature 🎉
> 
> > * which scoring use for the gridsearchcv (currently rmse) and metrics (currently mse , mae and r2)
> 
> I think rmse should suffice for now. I was also thinking it may be a good idea to allow the user the option to decide the scoring metric themselves.
> Ex:
> 
> ```python
> # In the case of classifier
> def fit(self, train_x, train_y, folds=3, scoring="accuracy"):
> ```
> We can definitely change this in a future PR.
> 
Yeah it's a good idea to let users choose. there is an option to use `make_scorer` in sklearn and they can also use that option too. need to be documented ofc.


> > * Is the param grid good enough or it can be improved
> 
> I definitely think it is good for now. It might be a good idea to get a solid workflow going before worrying about model coverage.

> 
> > * threshold and method to decide if tests are good enough - unlike in classifier where we can decide by accuracy in regression we need to measure it in another way like mse , rmse , mae and so on - can say that some tests are not reaching what I expected atm and mse can be high and not close to zero.
> 
> Yes, I like the idea of using a threshold for error and making sure the error score is under a certain value. Do you have an idea of what you think this threshold should be after implementing the class? On a side note, you will need to add the command for the regressor test cases here for them to be included in the ci pipeline:
> https://github.com/skekre98/simple_learn/blob/a5097b265d68bee1adba334fcefec10270bfe94c/scripts/custom.sh#L84-L87

I already added it to the CI 👍  
After testing it the class for few days I'm not sure on the treshold because different datasets give different values

Boston dataset :  

    
```
"Training Score": 4.624705269787787,

        "Mean Absolute Error": 1.663268136134209,

        "Mean Square Error": 6.4644087170138596,

        "R-Squared": 0.9234252226456211
```

    

diabetes dataset : 
  `  

        "Training Score": 54.7216547976174,

        "Mean Absolute Error": 43.59166030421379,

        "Mean Square Error": 2887.3116127762073,

        "R-Squared": 0.5130914574276866
    
`

I'm not sure yet if it's only the diabetes dataset that causes the problem (i'm failing to run other datasets from sklearn because they are too large ) and that's why I can't point on a good threshold.
I think the tests are sort of sanity check for the models and in classification with binary classification it's easier to get 95%

> 
> > * fail models implementation (might solve other issue #36 )
> 
> I have added a review comment regarding this. Let me know what you think. I really like the idea of saving the failed models as a class attribute as well.

I didn't see it in the review but you want to save the model class itself (in the code : RegressionClass ) ?

**Comment by skekre98** [comment]
> I already added it to the CI 👍
> After testing it the class for few days I'm not sure on the treshold because different datasets give different values

I can't seem to find it in the commit files. Are you planning on adding it after we finalize the threshold for the tests?

> I'm not sure yet if it's only the diabetes dataset that causes the problem (i'm failing to run other datasets from sklearn because they are too large ) and that's why I can't point on a good threshold.

No problem, should we just use a generic correlation test so we can at least provide some kind of robust validation for all regression datasets? For example we could use something like Levene's or Bartlett's test.

Levene:
```python
import numpy as np
from scipy.stats import levene

target = [8.88, 9.12, 9.04, 8.98, 9.00, 9.08, 9.01, 8.85, 9.06, 8.99]
error_predict = [8.88, 8.95, 9.29, 9.44, 9.15, 9.58, 8.36, 9.18, 8.67, 9.05]
accurate_predict = [8.95, 9.12, 8.95, 8.85, 9.03, 8.84, 9.07, 8.98, 8.86, 8.98]

stat, p = levene(target, error_predict)
print(p) # 0.009364737715584399

stat, p = levene(target, accurate_predict)
print(p) # 0.6275797623997227

var_array = [np.var(x, ddof=1) for x in [target, error_predict, accurate_predict]]
print(var_array) # [0.007054444444444413, 0.13073888888888888, 0.008890000000000002]
```

We could use a threshold like 0.05 to keep it generic and say that if the p-value if greater than 0.05 the model is accurate enough. WDYT?

> I didn't see it in the review but you want to save the model class itself (in the code : RegressionClass ) ?

Can you elaborate on this question, are you referring to #42?

**Comment by eranns** [comment]
> > I already added it to the CI 👍
> > After testing it the class for few days I'm not sure on the treshold because different datasets give different values
> 
> I can't seem to find it in the commit files. Are you planning on adding it after we finalize the threshold for the tests?
> 
Yes I didn't add it because tests are not reasonable enough because of the treshold.

> No problem, should we just use a generic correlation test so we can at least provide some kind of robust validation for all regression datasets? For example we could use something like Levene's or Bartlett's test.
> 
Yes it can make the tests more reliable.
> Levene:
> 
> ```python
> import numpy as np
> from scipy.stats import levene
> 
> target = [8.88, 9.12, 9.04, 8.98, 9.00, 9.08, 9.01, 8.85, 9.06, 8.99]
> error_predict = [8.88, 8.95, 9.29, 9.44, 9.15, 9.58, 8.36, 9.18, 8.67, 9.05]
> accurate_predict = [8.95, 9.12, 8.95, 8.85, 9.03, 8.84, 9.07, 8.98, 8.86, 8.98]
> 
> stat, p = levene(target, error_predict)
> print(p) # 0.009364737715584399
> 
> stat, p = levene(target, accurate_predict)
> print(p) # 0.6275797623997227
> 
> var_array = [np.var(x, ddof=1) for x in [target, error_predict, accurate_predict]]
> print(var_array) # [0.007054444444444413, 0.13073888888888888, 0.008890000000000002]
> ```
> 
> We could use a threshold like 0.05 to keep it generic and say that if the p-value if greater than 0.05 the model is accurate enough. WDYT?

alpha 0.05 is very common so it's good. 
I didn't understand what's the error predict in our case. we have x_train , y_train. then y_pred when we predict on train - we don't have three variables like this.

> 
> > I didn't see it in the review but you want to save the model class itself (in the code : RegressionClass ) ?
> 
> Can you elaborate on this question, are you referring to #42?
Well you said you want to save the model  as a class attribute but I only saved the name of the model

**Comment by skekre98** [comment]
> I didn't understand what's the error predict in our case. we have x_train , y_train. then y_pred when we predict on train - we don't have three variables like this.

Yes, in our case it will look something more along the lines of:
```python
train_x, train_y = load_dataset()
sr = SimpleRegressor()
sr.fit(train_x, train_y)
pred_y = sr.predict(train_x)
stat, p = levene(train_y, pred_y)
assertTrue(p > 0.05)
```

We will also need to add `scipy` into the list of requirements [here](https://github.com/skekre98/simple_learn/blob/f39cd243a6a6fce8c59ba2297b8170bafe9d21f0/setup.py#L28) and in requirements.txt

> Well you said you want to save the model as a class attribute but I only saved the name of the model

I can see you have also saved the actual model as `self.sk_model` which is what I was referring to. The class implementation looks good to me so all that's left is to see how it performs on the unit tests :)

File: simple_learn/regressors/simple_regressor.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+import numpy as np
+from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
+from sklearn.model_selection import GridSearchCV
+from sklearn.utils import all_estimators
+
+from simple_learn.regressors.param_grid import model_param_map
+
+
+class SimpleRegressor:
```
Comment (skekre98) [inline simple_learn/regressors/simple_regressor.py:33]: Would you also mind documenting the class prior to merge? See [SimpleClassifier](https://github.com/skekre98/simple_learn/blob/main/simple_learn/classifiers/simple_classifier.py) for reference. You can likely just copy a majority of the docs over and change a few terms.

Comment (eranns) [reply simple_learn/regressors/simple_regressor.py:33]: I will

**Comment by eranns** [comment]
> > I didn't understand what's the error predict in our case. we have x_train , y_train. then y_pred when we predict on train - we don't have three variables like this.
> 
> Yes, in our case it will look something more along the lines of:
> 
> ```python
> train_x, train_y = load_dataset()
> sr = SimpleRegressor()
> sr.fit(train_x, train_y)
> pred_y = sr.predict(train_x)
> stat, p = levene(train_y, pred_y)
> assertTrue(p > 0.05)
> ```
> 
👍 
> We will also need to add `scipy` into the list of requirements [here](https://github.com/skekre98/simple_learn/blob/f39cd243a6a6fce8c59ba2297b8170bafe9d21f0/setup.py#L28) and in requirements.txt
> 
> > Well you said you want to save the model as a class attribute but I only saved the name of the model
> 
> I can see you have also saved the actual model as `self.sk_model` which is what I was referring to. The class 
implementation looks good to me so all that's left is to see how it performs on the unit tests :)

the self.sk_model is something that was in classifier and it was also fine for a the regression class so I didnt touch it.
I will run the tests and send a merge request as soon as I can

**Comment by eranns** [comment]
Hi @skekre98 

I ran the tests in the last few days (I need to do it over night because else it wont finish :( )
While datasets from sklearn like boston passing, I saw that diabetes dataset is failing and the p value is close to zero.

I didn't see many works with the diabetes dataset so I used another dataset called california housing data which is larger than the others.
the p-value was `1.1978903766770936e-28` and the test failed but the score for the trainset seems alright : 

this is the selected model : 


```
{
    "Type": "GradientBoostingRegressor",

    "Training Duration": "12.884101867675781s",

    "GridSearch Duration": "11695.488729715347s",

    "Parameters": {

        "criterion": "friedman_mse",

        "learning_rate": 0.1,

        "loss": "huber",

        "max_depth": 6,

        "max_features": "auto",

        "n_estimators": 100
    },
  
  "Metrics": {
  
      "Training Score": 0.6315197740553223,

        "Mean Absolute Error": 0.2580415308858338,

        "Mean Square Error": 0.1597509137651601,

        "R-Squared": 0.8800263769567358
    }
}
```

It's hard to estimate and get intuition on how to test it when it takes to so long. let me know what do you think

**Comment by skekre98** [comment]
@eranns this looks good to me. If you'd like to leverage the CI for your testing, you can push your changes for the tests to this branch and you can see the logs for your commit in the [Actions](https://github.com/skekre98/simple_learn/actions) tab. 

So has the p-test not succeeded for any of the sklearn datasets? If so, perhaps we should try looking into a different form of validation.

**Comment by eranns** [comment]
> @eranns this looks good to me. If you'd like to leverage the CI for your testing, you can push your changes for the tests to this branch and you can see the logs for your commit in the [Actions](https://github.com/skekre98/simple_learn/actions) tab.

I wanted to use the CI for tests but if the regression not part of the CI tests now would they run it ? 
> 
> So has the p-test not succeeded for any of the sklearn datasets? If so, perhaps we should try looking into a different form of validation.

Ok I will explain myself better. I took all the regression datasets in sklearn and tested on them. Also used make_regression method (this create simple datasets) to test it.

make regression - mse less than 1 , p value greater than 0.5
boston dataset - mse less than 1 , p value greater than 0.5
diabetes dataset - both mse and p value fail (mse high and p value really low)
house pricing dataset - mse low and p value low.

So if we take the house pricing  - we fail on p value (bad) and  mse is 1.x which is not bad at all - like you said our method of validation is not correct because good results could be considered bad in the tests

**Comment by skekre98** [comment]
> I wanted to use the CI for tests but if the regression not part of the CI tests now would they run it ?

Yes, you'll need to create the tests and call them in the test script(The CI pipeline checks out the tests in the PR not main branch):
https://github.com/skekre98/simple_learn/blob/2269dca292ab219ddc50dd8905408245fbf9ac07/scripts/custom.sh#L85-L88

```python
function run_tests {
    python -W ignore -m unittest tests/classifiers/simple_classifier_tests.py -v
    python -W ignore -m unittest tests/classifiers/simple_classifier_list_tests.py -v
    python -W ignore -m unittest tests/regressors/simple_regressor_tests.py -v   
}
```

Cool, in that case let's go ahead and use make_regression and the boston dataset for the tests. It can be expanded later on down the road. Also I thought we had decided to use 0.05 for the p value threshold, do you think 0.5 would be better for validation?

**Comment by eranns** [comment]
> > I wanted to use the CI for tests but if the regression not part of the CI tests now would they run it ?
> 
> Yes, you'll need to create the tests and call them in the test script(The CI pipeline checks out the tests in the PR not main branch):

Thanks that what I was missing in my mind.

> https://github.com/skekre98/simple_learn/blob/2269dca292ab219ddc50dd8905408245fbf9ac07/scripts/custom.sh#L85-L88
> 
> ```python
> function run_tests {
>     python -W ignore -m unittest tests/classifiers/simple_classifier_tests.py -v
>     python -W ignore -m unittest tests/classifiers/simple_classifier_list_tests.py -v
>     python -W ignore -m unittest tests/regressors/simple_regressor_tests.py -v   
> }
> ```
> 
> Cool, in that case let's go ahead and use make_regression and the boston dataset for the tests. It can be expanded later on down the road. Also I thought we had decided to use 0.05 for the p value threshold, do you think 0.5 would be better for validation?

Yes i had typo its 0.05

I will upload tests shortly. Thanks for your help.

**Review by skekre98 — APPROVED** [review]
Really great work @eranns 🎉 . This was a key feature for the package! I'll go ahead and merge :)
